aclk/https_client.c

```diff
@@ -29,38 +21,59 @@ static const char *http_req_type_to_str(http_req_type_t req) {
     }
 }
 
-typedef struct {
-    enum http_parse_state state;
-    int content_length;
-    int http_code;
-} http_parse_ctx;
-
 #define HTTP_PARSE_CTX_INITIALIZER { .state = HTTP_PARSE_INITIAL, .content_length = -1, .http_code = 0 }
-static inline void http_parse_ctx_clear(http_parse_ctx *ctx) {
+void http_parse_ctx_create(http_parse_ctx *ctx)
+{
     ctx->state = HTTP_PARSE_INITIAL;
     ctx->content_length = -1;
     ctx->http_code = 0;
+    ctx->headers = c_rhash_new(0);
+    ctx->flags = HTTP_PARSE_FLAGS_DEFAULT;
+}
+
+void http_parse_ctx_destroy(http_parse_ctx *ctx)
+{
+    c_rhash_iter_t iter;
+    const char *key;
+
+    c_rhash_iter_t_initialize(&iter);
+    while ( !c_rhash_iter_str_keys(ctx->headers, &iter, &key) ) {
+        void *val;
+        c_rhash_get_ptr_by_str(ctx->headers, key, &val);
+        freez(val);
+    }
+
+    c_rhash_destroy(ctx->headers);
 }
 
 #define POLL_TO_MS 100
 
-#define NEED_MORE_DATA  0
-#define PARSE_SUCCESS   1
-#define PARSE_ERROR    -1
 #define HTTP_LINE_TERM "\x0D\x0A"
 #define RESP_PROTO "HTTP/1.1 "
 #define HTTP_KEYVAL_SEPARATOR ": "
 #define HTTP_HDR_BUFFER_SIZE 256
 #define PORT_STR_MAX_BYTES 12
 
-static void process_http_hdr(http_parse_ctx *parse_ctx, const char *key, const char *val)
+static void process_http_hdr(http_parse_ctx *ctx, const char *key, const char *val)
 {
     // currently we care only about content-length
     // but in future the way this is written
     // it can be extended
-    if (!strcmp("content-length", key)) {
-        parse_ctx->content_length = atoi(val);
+    if (ctx->content_length < 0 && !strcmp("content-length", key)) {
+        ctx->content_length = atoi(val);
```

Comment:
I suggest to use our `str2i` instead `atoi`.

Reply:
str2u

Reply:
str4everyone